File: app/services/snapshot_service.py
from __future__ import annotations

from decimal import Decimal

from sqlalchemy import func, select
from sqlalchemy.orm import Session, selectinload

from app.audit import record_audit
from app.models import (
    BomLine,
    CalculationSnapshot,
    EmissionFactor,
    EnergyActivity,
    FactorVersion,
    MappingStatus,
    MaterialProcessMapping,
    ProcessRoute,
    Product,
    ProductVersion,
    RouteStep,
    TransportActivity,
)
from app.modules.catalog.units import UnitConversionError, convert_activity_amount
from app.modules.snapshots.contracts import SnapshotPayload
from app.utils import hash_payload


class SnapshotValidationError(ValueError):
    def __init__(self, errors: list[dict]):
        super().__init__("Snapshot validation failed")
        self.errors = errors


class SnapshotService:
    def __init__(self, db: Session):
        self.db = db

    def create(
        self,
        *,
        sku: str,
        factor_set_version: str,
        route_version: str,
        boundary: str,
        actor: str,
    ) -> CalculationSnapshot:
        product = self.db.scalar(select(Product).where(Product.sku == sku))
        if not product:
            raise ValueError(f"Unknown product: {sku}")
        product_version = self.db.scalar(
            select(ProductVersion)
            .where(ProductVersion.product_id == product.id)
            .order_by(ProductVersion.version.desc())
        )
        route = self.db.scalar(
            select(ProcessRoute)
            .options(selectinload(ProcessRoute.steps))
            .where(ProcessRoute.product_id == product.id, ProcessRoute.version == route_version)
        )
        bom_lines = list(
            self.db.scalars(
                select(BomLine)
                .options(
                    selectinload(BomLine.material),
                    selectinload(BomLine.factor_version).selectinload(FactorVersion.factor),
                )
                .where(BomLine.product_version_id == product_version.id)
                .order_by(BomLine.line_no)
            )
        )
        errors: list[dict] = []
        if not bom_lines:
            errors.append({"code": "BOM_MISSING", "message": "Product has no BOM lines"})
        if not route:
            errors.append({"code": "ROUTE_MISSING", "message": f"Route {route_version} not found"})
        elif not route.approved:
            errors.append({"code": "ROUTE_NOT_APPROVED", "message": f"Route {route_version} is not approved"})

        bom_payload = []
        stage_estimates = {
            "raw_materials": Decimal("0"),
            "inbound_transport": Decimal("0"),
            "manufacturing": Decimal("0"),
            "packaging": Decimal("0"),
        }
        openlca_parameters: dict[str, str] = {}
        for line in bom_lines:
            factor = line.factor_version
            if factor is None:
                errors.append(
                    {"code": "FACTOR_MISSING", "line": line.line_no, "material": line.material.code}
                )
                continue
            if not factor.approved:
                errors.append(
                    {"code": "FACTOR_NOT_APPROVED", "line": line.line_no, "factor": factor.factor.factor_code}
                )
            if line.weight_kg_each is None:
                errors.append({"code": "BOM_WEIGHT_MISSING", "line": line.line_no})
                continue
            mass_kg = Decimal(line.quantity) * Decimal(line.weight_kg_each)
            try:
                activity_amount = convert_activity_amount(
                    quantity=Decimal(line.quantity),
                    weight_kg_each=Decimal(line.weight_kg_each),
                    activity_unit=factor.activity_unit,
                    density_kg_m3=(
                        Decimal(factor.density_kg_m3) if factor.density_kg_m3 else None
                    ),
                )
            except UnitConversionError:
                code = (
                    "DENSITY_MISSING"
                    if factor.activity_unit.lower() in {"m³", "m3"}
                    else "UNIT_CONVERSION_MISSING"
                )
                errors.append(
                    {
                        "code": code,
                        "line": line.line_no,
                        "from": line.unit,
                        "to": factor.activity_unit,
                        "material": line.material.code,
                    }
                )
                continue
            mapping = self.db.scalar(
                select(MaterialProcessMapping)
                .where(
                    MaterialProcessMapping.material_id == line.material_id,
                    MaterialProcessMapping.status == MappingStatus.approved,
                )
                .order_by(MaterialProcessMapping.created_at.desc())
            )
            if not mapping:
                errors.append(
                    {
                        "code": "OPENLCA_MAPPING_MISSING",
                        "line": line.line_no,
                        "material": line.material.code,
                    }
                )
            contribution = activity_amount * Decimal(factor.value)
            stage_estimates[line.stage] += contribution
            parameter_name = f"mat_{line.material.code.lower().replace('-', '_')}"
            openlca_parameters[parameter_name] = str(activity_amount)
            bom_payload.append(
                {
                    "line_no": line.line_no,
                    "material_code": line.material.code,
                    "part_name": line.part_name,
                    "stage": line.stage,
                    "quantity": str(line.quantity),
                    "unit": line.unit,
                    "weight_kg_each": str(line.weight_kg_each),
                    "mass_kg": str(mass_kg),
                    "factor_code": factor.factor.factor_code,
                    "factor_version": factor.version,
                    "factor_value": str(factor.value),
                    "factor_activity_unit": factor.activity_unit,
                    "factor_source": factor.source,
                    "activity_amount": str(activity_amount),
                    "mapping": {
                        "process_uuid": mapping.process_uuid,
                        "reference_flow_uuid": mapping.reference_flow_uuid,
                        "openlca_unit": mapping.openlca_unit,
                        "database_version": mapping.database_version,
                    }
                    if mapping
                    else None,
                }
            )

        energy_payload = self._energy_payload(product_version, route, errors, stage_estimates, openlca_parameters)
        transport_payload = self._transport_payload(product_version, errors, stage_estimates, openlca_parameters)
        payload_model = SnapshotPayload(
            schema_version=1,
            sku=product.sku,
            product_name=product.name,
            functional_unit=f"1 completed and packaged {product.name} at factory gate",
            boundary=boundary,
            factor_set_version=factor_set_version,
            product_version=product_version.version,
            route_version=route_version,
            bom=bom_payload,
            energy=energy_payload,
            transport=transport_payload,
            openlca_parameters=openlca_parameters,
            stage_estimates={key: str(value) for key, value in stage_estimates.items()},
        )
        payload = payload_model.model_dump(mode="json")
        if errors:
            raise SnapshotValidationError(errors)
        version = int(
            self.db.scalar(
                select(func.max(CalculationSnapshot.version)).where(
                    CalculationSnapshot.product_id == product.id
                )
            )
            or 0
        ) + 1
        snapshot = CalculationSnapshot(
            product_id=product.id,
            version=version,
            product_version_id=product_version.id,
            route_id=route.id,
            factor_set_version=factor_set_version,
            boundary=boundary,
            payload=payload,
            manifest_hash=hash_payload(payload),
            validation_errors=[],
            created_by=actor,
        )
        self.db.add(snapshot)
        self.db.flush()
        record_audit(
            self.db,
            actor=actor,
            action="snapshot.created",
            object_type="calculation_snapshot",
            object_id=snapshot.id,
            after_hash=snapshot.manifest_hash,
            details={"sku": sku, "version": version},
        )
        self.db.flush()
        return snapshot

    def _energy_payload(
        self,
        product_version: ProductVersion,
        route: ProcessRoute | None,
        errors: list[dict],
        stage_estimates: dict[str, Decimal],
        openlca_parameters: dict[str, str],
    ) -> list[dict]:
        activities = list(
            self.db.scalars(
                select(EnergyActivity).where(EnergyActivity.product_version_id == product_version.id)
            )
        )
        payload = []
        if activities:
            source_items = activities
        else:
            if not route:
                return []
            grid_factor = self.db.scalar(
                select(FactorVersion)
                .join(EmissionFactor)
                .where(
                    FactorVersion.approved.is_(True),
                    EmissionFactor.name.ilike("%Grid Electricity%"),
                )
                .order_by(FactorVersion.version.desc())
            ) or self.db.scalar(
                select(FactorVersion)
                .join(EmissionFactor)
                .where(
                    FactorVersion.approved.is_(True),
                    EmissionFactor.name.ilike("%市电%"),
                )
                .order_by(FactorVersion.version.desc())
            )
            if not grid_factor:
                errors.append({"code": "GRID_FACTOR_MISSING", "message": "Approved grid factor is required"})
                return []
            source_items = [
                {
                    "process_code": step.process_code,
                    "name": step.name,
                    "amount": step.energy_kwh_per_unit or Decimal("0"),
                    "unit": "kWh",
                    "factor": grid_factor,
                    "source": "approved route standard",
                }
                for step in sorted(route.steps, key=lambda x: x.sequence)
            ]
        for item in source_items:
            if isinstance(item, dict):
                factor = item["factor"]
                process_code = item["process_code"]
                name = item["name"]
                amount = Decimal(item["amount"])
                source = item["source"]
            else:
                factor = self.db.get(FactorVersion, item.factor_version_id)
                step = self.db.get(RouteStep, item.route_step_id) if item.route_step_id else None
                process_code = step.process_code if step else "ENERGY"
                name = step.name if step else item.energy_type
                amount = Decimal(item.amount)
                source = item.source
                if not item.approved:
                    errors.append({"code": "ENERGY_ACTIVITY_NOT_APPROVED", "id": item.id})
            if not factor or not factor.approved:
                errors.append({"code": "ENERGY_FACTOR_NOT_APPROVED", "process": process_code})
                continue
            contribution = amount * Decimal(factor.value)
            stage_estimates["manufacturing"] += contribution
            parameter_name = f"energy_{process_code.lower().replace('-', '_')}"
            openlca_parameters[parameter_name] = str(amount)
            payload.append(
                {
                    "process_code": process_code,
                    "name": name,
                    "amount": str(amount),
                    "unit": "kWh",
                    "factor_code": factor.factor.factor_code,
                    "factor_value": str(factor.value),
                    "source": source,
                }
            )
        return payload

    def _transport_payload(
        self,
        product_version: ProductVersion,
        errors: list[dict],
        stage_estimates: dict[str, Decimal],
        openlca_parameters: dict[str, str],
    ) -> list[dict]:
        activities = list(
            self.db.scalars(
                select(TransportActivity).where(
                    TransportActivity.product_version_id == product_version.id
                )
            )
        )
        if not activities:
            errors.append(
                {
                    "code": "INBOUND_TRANSPORT_MISSING",
                    "message": "At least one approved inbound transport activity is required",
                }
            )
            return []
        payload = []
        for index, item in enumerate(activities, start=1):
            factor = self.db.get(FactorVersion, item.factor_version_id)
            if not item.approved:
                errors.append({"code": "TRANSPORT_NOT_APPROVED", "id": item.id})
            if not factor or not factor.approved:
                errors.append({"code": "TRANSPORT_FACTOR_NOT_APPROVED", "id": item.id})
                continue
            tkm = Decimal(item.mass_kg) / Decimal("1000") * Decimal(item.distance_km)
            stage_estimates["inbound_transport"] += tkm * Decimal(factor.value)
            openlca_parameters[f"transport_{index}_tkm"] = str(tkm)
            payload.append(
                {
                    "mode": item.mode,
                    "distance_km": str(item.distance_km),
                    "mass_kg": str(item.mass_kg),
                    "load_factor": str(item.load_factor) if item.load_factor else None,
                    "factor_code": factor.factor.factor_code,
                    "factor_value": str(factor.value),
                    "material_code": None,
                    "source": item.source,
                }
            )
        return payload

File: java-platform/src/main/java/com/airpaq/pcf/snapshots/SnapshotService.java
package com.airpaq.pcf.snapshots;

import com.airpaq.pcf.audit.AuditService;
import com.airpaq.pcf.infrastructure.api.ApiContracts.SnapshotCreate;
import com.airpaq.pcf.infrastructure.api.ApiContracts.SnapshotOut;
import com.airpaq.pcf.infrastructure.json.CanonicalJson;
import com.airpaq.pcf.infrastructure.web.DomainException;
import java.math.BigDecimal;
import java.time.Clock;
import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class SnapshotService {

    private static final BigDecimal ZERO = BigDecimal.ZERO;

    private final JdbcTemplate jdbc;
    private final CanonicalJson canonicalJson;
    private final AuditService audit;
    private final Clock clock = Clock.systemUTC();

    public SnapshotService(JdbcTemplate jdbc, CanonicalJson canonicalJson, AuditService audit) {
        this.jdbc = jdbc;
        this.canonicalJson = canonicalJson;
        this.audit = audit;
    }

    @Transactional
    public SnapshotOut create(String sku, SnapshotCreate request, String actor) {
        var product =
                one(
                        """
                        select id, sku, name from product where sku = ?
                        """,
                        sku);
        if (product == null) throw DomainException.notFound("Unknown product: " + sku);
        var productVersion =
                one(
                        """
                        select id, version from product_version
                        where product_id = ? order by version desc limit 1
                        """,
                        product.get("id"));
        if (productVersion == null) {
            throw DomainException.unprocessable("Product has no version");
        }
        var route =
                one(
                        """
                        select id, approved from process_route
                        where product_id = ? and version = ?
                        """,
                        product.get("id"),
                        request.routeVersion());
        var errors = new ArrayList<Map<String, Object>>();
        if (route == null) {
            errors.add(
                    Map.of(
                            "code",
                            "ROUTE_MISSING",
                            "message",
                            "Route " + request.routeVersion() + " not found"));
        } else if (!Boolean.TRUE.equals(route.get("approved"))) {
            errors.add(
                    Map.of(
                            "code",
                            "ROUTE_NOT_APPROVED",
                            "message",
                            "Route " + request.routeVersion() + " is not approved"));
        }

        var stages = stageMap();
        var parameters = new LinkedHashMap<String, String>();
        var bomPayload = new ArrayList<Map<String, Object>>();
        var bom =
                jdbc.queryForList(
                        """
                        select b.line_no, b.material_id, b.part_name, b.quantity, b.unit,
                               b.weight_kg_each, b.stage, m.code material_code,
                               f.factor_code, fv.version factor_version, fv.value factor_value,
                               fv.activity_unit, fv.source factor_source, fv.approved factor_approved,
                               fv.density_kg_m3,
                               mp.process_uuid, mp.reference_flow_uuid, mp.openlca_unit,
                               mp.database_version
                        from bom_line b
                        join material m on m.id = b.material_id
                        left join factor_version fv on fv.id = b.factor_version_id
                        left join emission_factor f on f.id = fv.factor_id
                        left join lateral (
                          select * from material_process_mapping x
                          where x.material_id = b.material_id and x.status = 'approved'
                          order by x.created_at desc limit 1
                        ) mp on true
                        where b.product_version_id = ?
                        order by b.line_no
                        """,
                        productVersion.get("id"));
        if (bom.isEmpty()) {
            errors.add(Map.of("code", "BOM_MISSING", "message", "Product has no BOM lines"));
        }
        for (var line : bom) {
            var lineNo = ((Number) line.get("line_no")).intValue();
            if (line.get("factor_value") == null) {
                errors.add(
                        Map.of(
                                "code",
                                "FACTOR_MISSING",
                                "line",
                                lineNo,
                                "material",
                                line.get("material_code")));
                continue;
            }
            if (!Boolean.TRUE.equals(line.get("factor_approved"))) {
                errors.add(
                        Map.of(
                                "code",
                                "FACTOR_NOT_APPROVED",
                                "line",
                                lineNo,
                                "factor",
                                line.get("factor_code")));
            }
            if (line.get("weight_kg_each") == null) {
                errors.add(Map.of("code", "BOM_WEIGHT_MISSING", "line", lineNo));
                continue;
            }
            var quantity = decimal(line.get("quantity"));
            var weight = decimal(line.get("weight_kg_each"));
            var mass = quantity.multiply(weight);
            var unit = String.valueOf(line.get("activity_unit"));
            BigDecimal activity;
            if (unit.equalsIgnoreCase("kg") || unit.equalsIgnoreCase("kilogram")) {
                activity = mass;
            } else if (unit.equalsIgnoreCase("m3") || unit.equalsIgnoreCase("m³")) {
                if (line.get("density_kg_m3") == null) {
                    errors.add(
                            Map.of(
                                    "code",
                                    "DENSITY_MISSING",
                                    "line",
                                    lineNo,
                                    "material",
                                    line.get("material_code")));
                    continue;
                }
                activity =
                        mass.divide(
                                        decimal(line.get("density_kg_m3")),
                                        12,
                                        java.math.RoundingMode.HALF_UP)
                                .stripTrailingZeros();
            } else {
                errors.add(
                        Map.of(
                                "code",
                                "UNIT_CONVERSION_MISSING",
                                "line",
                                lineNo,
                                "from",
                                line.get("unit"),
                                "to",
                                unit));
                continue;
            }
            if (line.get("process_uuid") == null) {
                errors.add(
                        Map.of(
                                "code",
                                "OPENLCA_MAPPING_MISSING",
                                "line",
                                lineNo,
                                "material",
                                line.get("material_code")));
            }
            var stage = String.valueOf(line.get("stage"));
            stages.merge(
                    stage, activity.multiply(decimal(line.get("factor_value"))), BigDecimal::add);
            var parameterName =
                    "mat_"
                            + String.valueOf(line.get("material_code"))
                                    .toLowerCase()
                                    .replace("-", "_");
            parameters.put(parameterName, plain(activity));
            var payload = new LinkedHashMap<String, Object>();
            payload.put("line_no", lineNo);
            payload.put("material_code", line.get("material_code"));
            payload.put("part_name", line.get("part_name"));
            payload.put("stage", stage);
            payload.put("quantity", plain(quantity));
            payload.put("unit", line.get("unit"));
            payload.put("weight_kg_each", plain(weight));
            payload.put("mass_kg", plain(mass));
            payload.put("factor_code", line.get("factor_code"));
            payload.put("factor_version", line.get("factor_version"));
            payload.put("factor_value", plain(decimal(line.get("factor_value"))));
            payload.put("factor_activity_unit", unit);
            payload.put("factor_source", line.get("factor_source"));
            payload.put("activity_amount", plain(activity));
            payload.put(
                    "mapping",
                    line.get("process_uuid") == null
                            ? null
                            : Map.of(
                                    "process_uuid", line.get("process_uuid"),
                                    "reference_flow_uuid", line.get("reference_flow_uuid"),
                                    "openlca_unit", line.get("openlca_unit"),
                                    "database_version", line.get("database_version")));
            bomPayload.add(payload);
        }

        var energyPayload =
                energyPayload(
                        String.valueOf(productVersion.get("id")),
                        route == null ? null : String.valueOf(route.get("id")),
                        errors,
                        stages,
                        parameters);
        var transportPayload =
                transportPayload(
                        String.valueOf(productVersion.get("id")), errors, stages, parameters);
        if (!errors.isEmpty()) {
            throw DomainException.unprocessable(
                    Map.of("code", "SNAPSHOT_INVALID", "errors", errors));
        }
        var payload = new LinkedHashMap<String, Object>();
        payload.put("schema_version", 1);
        payload.put("sku", sku);
        payload.put("product_name", product.get("name"));
        payload.put(
                "functional_unit",
                "1 completed and packaged " + product.get("name") + " at factory gate");
        payload.put("boundary", request.boundary());
        payload.put("factor_set_version", request.factorSetVersion());
        payload.put("product_version", productVersion.get("version"));
        payload.put("route_version", request.routeVersion());
        payload.put("bom", bomPayload);
        payload.put("energy", energyPayload);
        payload.put("transport", transportPayload);
        payload.put("openlca_parameters", parameters);
        payload.put(
                "stage_estimates",
                stages.entrySet().stream()
                        .collect(
                                java.util.stream.Collectors.toMap(
                                        Map.Entry::getKey,
                                        entry -> plain(entry.getValue()),
                                        (left, right) -> right,
                                        LinkedHashMap::new)));
        var hash = canonicalJson.sha256(payload);
        var version =
                jdbc.queryForObject(
                        "select coalesce(max(version), 0) + 1 from calculation_snapshot where product_id = ?",
                        Integer.class,
                        product.get("id"));
        var id = UUID.randomUUID().toString();
        var now = java.sql.Timestamp.from(Instant.now(clock));
        jdbc.update(
                """
                insert into calculation_snapshot
                  (id, product_id, version, product_version_id, route_id,
                   factor_set_version, boundary, payload, manifest_hash,
                   validation_errors, created_by, created_at)
                values (?, ?, ?, ?, ?, ?, ?, ?::json, ?, '[]'::json, ?, ?)
                """,
                id,
                product.get("id"),
                version,
                productVersion.get("id"),
                route.get("id"),
                request.factorSetVersion(),
                request.boundary(),
                canonicalJson.string(payload),
                hash,
                actor,
                now);
        audit.record(
                actor,
                "snapshot.created",
                "calculation_snapshot",
                id,
                null,
                hash,
                Map.of("sku", sku, "version", version));
        return new SnapshotOut(
                id,
                version,
                request.factorSetVersion(),
                request.boundary(),
                hash,
                List.of(),
                now.toInstant());
    }

    private List<Map<String, Object>> energyPayload(
            String productVersionId,
            String routeId,
            List<Map<String, Object>> errors,
            Map<String, BigDecimal> stages,
            Map<String, String> parameters) {
        var activities =
                jdbc.queryForList(
                        """
                        select ea.*, rs.process_code, rs.name, fv.value factor_value,
                               f.factor_code
                        from energy_activity ea
                        left join route_step rs on rs.id = ea.route_step_id
                        left join factor_version fv on fv.id = ea.factor_version_id
                        left join emission_factor f on f.id = fv.factor_id
                        where ea.product_version_id = ?
                        """,
                        productVersionId);
        if (activities.isEmpty() && routeId != null) {
            var factor =
                    one(
                            """
                            select fv.id factor_version_id, fv.value factor_value,
                                   f.factor_code
                            from factor_version fv join emission_factor f on f.id = fv.factor_id
                            where fv.approved = true
                              and (f.name ilike '%Grid Electricity%' or f.name ilike '%市电%')
                            order by fv.version desc limit 1
                            """);
            if (factor == null) {
                errors.add(
                        Map.of(
                                "code", "GRID_FACTOR_MISSING",
                                "message", "Approved grid factor is required"));
                return List.of();
            }
            activities =
                    jdbc.queryForList(
                            """
                            select process_code, name, coalesce(energy_kwh_per_unit, 0) amount,
                                   'approved route standard' source
                            from route_step where route_id = ? order by sequence
                            """,
                            routeId);
            for (var item : activities) {
                item.putAll(factor);
                item.put("approved", true);
            }
        }
        var payload = new ArrayList<Map<String, Object>>();
        for (var item : activities) {
            if (item.containsKey("approved") && !Boolean.TRUE.equals(item.get("approved"))) {
                errors.add(Map.of("code", "ENERGY_ACTIVITY_NOT_APPROVED", "id", item.get("id")));
            }
            if (item.get("factor_value") == null) {
                errors.add(
                        Map.of(
                                "code",
                                "ENERGY_FACTOR_NOT_APPROVED",
                                "process",
                                item.get("process_code")));
                continue;
            }
            var amount = decimal(item.get("amount"));
            stages.merge(
                    "manufacturing",
                    amount.multiply(decimal(item.get("factor_value"))),
                    BigDecimal::add);
            var code = String.valueOf(item.getOrDefault("process_code", "ENERGY"));
            parameters.put("energy_" + code.toLowerCase().replace("-", "_"), plain(amount));
            payload.add(
                    Map.of(
                            "process_code", code,
                            "name",
                                    item.getOrDefault(
                                            "name", item.getOrDefault("energy_type", "Energy")),
                            "amount", plain(amount),
                            "unit", "kWh",
                            "factor_code", item.get("factor_code"),
                            "factor_value", plain(decimal(item.get("factor_value"))),
                            "source", item.get("source")));
        }
        return payload;
    }

    private List<Map<String, Object>> transportPayload(
            String productVersionId,
            List<Map<String, Object>> errors,
            Map<String, BigDecimal> stages,
            Map<String, String> parameters) {
        var rows =
                jdbc.queryForList(
                        """
                        select t.*, fv.value factor_value, fv.approved factor_approved,
                               f.factor_code
                        from transport_activity t
                        left join factor_version fv on fv.id = t.factor_version_id
                        left join emission_factor f on f.id = fv.factor_id
                        where t.product_version_id = ?
                        """,
                        productVersionId);
        if (rows.isEmpty()) {
            errors.add(
                    Map.of(
                            "code", "INBOUND_TRANSPORT_MISSING",
                            "message",
                                    "At least one approved inbound transport activity is required"));
            return List.of();
        }
        var payload = new ArrayList<Map<String, Object>>();
        var index = 0;
        for (var item : rows) {
            index++;
            if (!Boolean.TRUE.equals(item.get("approved"))) {
                errors.add(Map.of("code", "TRANSPORT_NOT_APPROVED", "id", item.get("id")));
            }
            if (!Boolean.TRUE.equals(item.get("factor_approved"))) {
                errors.add(Map.of("code", "TRANSPORT_FACTOR_NOT_APPROVED", "id", item.get("id")));
                continue;
            }
            var distance = decimal(item.get("distance_km"));
            var mass = decimal(item.get("mass_kg"));
            var tkm = mass.divide(new BigDecimal("1000")).multiply(distance);
            stages.merge(
                    "inbound_transport",
                    tkm.multiply(decimal(item.get("factor_value"))),
                    BigDecimal::add);
            parameters.put("transport_" + index + "_tkm", plain(tkm));
            var row = new LinkedHashMap<String, Object>();
            row.put("mode", item.get("mode"));
            row.put("distance_km", plain(distance));
            row.put("mass_kg", plain(mass));
            row.put(
                    "load_factor",
                    item.get("load_factor") == null
                            ? null
                            : plain(decimal(item.get("load_factor"))));
            row.put("factor_code", item.get("factor_code"));
            row.put("factor_value", plain(decimal(item.get("factor_value"))));
            row.put("material_code", null);
            row.put("source", item.get("source"));
            payload.add(row);
        }
        return payload;
    }

    private Map<String, Object> one(String sql, Object... args) {
        var rows = jdbc.queryForList(sql, args);
        return rows.isEmpty() ? null : rows.getFirst();
    }

    private static LinkedHashMap<String, BigDecimal> stageMap() {
        var stages = new LinkedHashMap<String, BigDecimal>();
        stages.put("raw_materials", ZERO);
        stages.put("inbound_transport", ZERO);
        stages.put("manufacturing", ZERO);
        stages.put("packaging", ZERO);
        return stages;
    }

    private static BigDecimal decimal(Object value) {
        return value instanceof BigDecimal decimal
                ? decimal
                : new BigDecimal(String.valueOf(value));
    }

    private static String plain(BigDecimal value) {
        return value.stripTrailingZeros().toPlainString();
    }
}
